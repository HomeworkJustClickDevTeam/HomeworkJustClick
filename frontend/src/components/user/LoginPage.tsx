import React, {ChangeEvent, useContext, useState} from "react"
import {loginPostgresService} from "../../services/postgresDatabaseServices"
import {Link, useNavigate} from "react-router-dom"
import HomePageContext from "../../contexts/HomePageContext"
import login_left_circle from './login_left_circle.svg';
import login_right_circle from './login_right_circle.svg';
import {LoginUserInterface} from "../../types/LoginUserInterface";
import HomePage from "../home/HomePage";


const LoginPage = () => {
  const [user, setUser] = useState<LoginUserInterface>({
    email: "",
    password: "",
  })
  const {setHomePageIn} = useContext(HomePageContext)
  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault()

    try {
      const response = await loginPostgresService(user)
      if (response?.data.token) {
        localStorage.setItem("user", JSON.stringify(response.data))
        setHomePageIn(true)
        return <HomePage/>
      } else {
        console.log("Zle haslo / uzytkownik")
      }
    } catch (e) {
      console.log(e)
    }
  }



  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target
    setUser((prevState) => ({
      ...prevState,
      [name]: value,
    }))
  }

  return (
      <div className='flex w-screen flex-col items-center text-center font-lato text-sm select-none'>
        <img className="fixed left-[7%]  bottom-[3%] scale-50 xl:scale-100 -z-50" src = {login_left_circle} alt="Kółko po lewej stronie"></img>
        <img className="fixed right-[9%] top-[18%] scale-50 translate-x-[25%] xl:transform-none -z-50" src= {login_right_circle} alt="Kółko po prawej stronie"></img>
        <h1 className='mt-32 mb-16 text-6xl'>Zaloguj się</h1>
        <form onSubmit={handleSubmit} className='flex flex-col items-center'>
          <input
              name="email"
              type="email"
              placeholder="Email"
              onChange={handleChange}
              className='border-b-2 border-b-light_gray mb-8 text-center placeholder:text-light_gray placeholder:text-[12px] w-60 xl:text-[20px] xl:placeholder:text-[20px] xl:w-[320px]'
          />
          <input
              name="password"
              type="password"
              placeholder="Hasło"
              onChange={handleChange}
              className='border-b-2 border-b-light_gray mb-12 text-center placeholder:text-light_gray placeholder:text-[12px] w-60 xl:text-[20px] xl:placeholder:text-[20px] xl:w-[320px]'
          />
          <button type="submit" className=' border-main_blue w-40 h-8 rounded-[5px] border-2 xl:w-[288px] xl:h-[56px] xl:border-[3px] xl:rounded-[10px] text-main_blue xl:text-[28px] hover:bg-hover_gray hover:shadow-md active:bg-opacity-60'>Zaloguj się</button>
          <p className='text-[8px] xl:text-base text-main_blue'>Nie pamiętasz hasła?</p>
          <p className='text-[8px] xl:text-base mt-4'>Nie masz jeszcze konta?</p>
          <Link to="/register">
            <button type="button" className=' bg-main_blue w-40 h-8 rounded-[5px] xl:rounded-[10px] xl:w-[288px] xl:h-[56px] text-white xl:text-[28px] hover:bg-hover_blue hover:shadow-md active:shadow-none'>Zarejestruj się</button>
          </Link>
        </form>
      </div>
  )
}
export default LoginPage
