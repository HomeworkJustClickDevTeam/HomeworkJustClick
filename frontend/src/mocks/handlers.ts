import { http, HttpResponse, delay } from "msw";
import { be } from "date-fns/locale";


export const homePageHandlers = [
  http.get('http://localhost:8080/api/groups/byTeacher/:id', async ({params}) => {
    await delay(100)
    if(params.id !== "undefined" && params.id !== "null" && params.id === '0'){
      return HttpResponse.json(null, {status: 404})
    }
    return HttpResponse.json([
        {
          id: 0,
          name: "Example Group 1",
          description: "Example desc",
          color: 0,
          archived: false
        },
        {
          id: 2,
          name: "Example Group 2",
          description: "Example desc 2",
          color: 1,
          archived: false
        }
      ],
      {
        status: 200
      })
  }),
  http.get('http://localhost:8080/api/groups/byStudent/:id', async ({params}) => {
    await delay(100)
    return HttpResponse.json(null, {status: 404})
  }),
  http.get('http://localhost:8080/api/groups/byUser/:id', async ({params}) => {
    const {id} = params
    await delay(100)
    if(params.id !== "undefined" && params.id !== "null" && params.id === '0'){
      return HttpResponse.json(null, {status: 404})
    }
    return HttpResponse.json( [
          {
            id: 0,
            name: "Example Group 1",
            description: "Example desc",
            color: 0,
            archived: false
          },
          {
            id: 2,
            name: "Example Group 2",
            description: "Example desc 2",
            color: 1,
            archived: false
          }
        ],
      {
        status: 200
      })
  }),
  http.get('http://localhost:8080/api/groups/userCheckWithRole/:userId/:groupId', async ({params}) => {
    const {userId, groupId} = params
    await delay(100)
    if(userId === '3'){
      return new Response("Teacher")
    }
    else
      return new Response("Student")
  })
]

export const loginPageHandler = [
  http.post('http://localhost:8080/api/auth/authenticate', async () => {

    await delay(100)
    return HttpResponse.json({
        token: "eyJhbGc3ZGEuY29tIiwiaWF0IjoxNzAwOTkwNjkyLCJleHAiOjE3MDA5OTQyOTJ9.B22HfZ1nIWfmHB66lt0vppGE86eERfTM-f61GG--v0LB-7dulmb4H11Q3niO4NYSJncEE5JzRhoBHhdraIP3og",
        id: 1,
        role: "USER",
        message: "ok",
        color: 0,
        name: "Jan",
        lastname: "Kowalski",
        index: 123123
      }, {
      status: 200
    })
  })
]