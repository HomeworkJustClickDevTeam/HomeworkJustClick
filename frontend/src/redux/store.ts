import {combineReducers, configureStore, PreloadedState} from "@reduxjs/toolkit"
import userStateReducer from "./userStateSlice"
import roleReducer from "./roleSlice"
import groupReducer from "./groupSlice"
import isLoadingReducer from "./isLoadingSlice"
import homePageInReducer from "./homePageInSlice"



const rootReducer = combineReducers({
  userState: userStateReducer,
  role: roleReducer,
  group: groupReducer,
  isLoading: isLoadingReducer,
  homePageIn: homePageInReducer
})
export const setupStore = (preloadedState?: PreloadedState<RootState>) =>{
  return configureStore({
    reducer: rootReducer,
    preloadedState
  })
}


export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']