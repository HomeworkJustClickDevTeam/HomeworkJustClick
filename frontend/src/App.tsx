import React, { useEffect, useState } from "react"

import './App.css'
// import "./assets/App.css"
import RegisterPage from "./components/user/RegisterPage"
import LoginPage from "./components/user/LoginPage"
import HomePage from "./components/home/HomePage"
import { BrowserRouter, Route, Routes } from "react-router-dom"
import HomeGuestPage from "./components/home/HomeGuestPage"
import GroupCreatePage from "./components/group/GroupCreatePage"
import GroupPage from "./components/group/GroupPage"
import UserContext from "./UserContext"
import AssignmentsGroupDisplayedPage from "./components/assigments/AssignmentsGroupDisplayedPage"
import AddAssigmentPage from "./components/assigments/AddAssigmentPage"
import AssigmentSpecPage from "./components/assigments/AssigmentSpecPage"
import { useImmerReducer } from "use-immer"
import DispatchContext from "./DispatchContext"
import Header from "./components/header/Header"
import GroupUsersPage from "./components/group/GroupUsersPage"
import NotFoundPage from "./components/errors/NotFoundPage"
import GroupRoleContext from "./GroupRoleContext"
import GroupSetRoleContext from "./GroupSetRoleContext"
import AssignmentsTypesPage from "./components/assigments/AssignmentsTypesPage"
import AssignmentsStudentDisplayedPage from "./components/assigments/AssignmentsStudentDisplayedPage"
import SolutionsTypesPage from "./components/solution/SolutionsTypesPage"
import SolutionPage from "./components/solution/SolutionPage"
import UserSettingsPage from "./components/user/UserSettingsPage"
import UserGeneralSettingsPage from "./components/user/UserGeneralSettingsPage"
import UserSecuritySettingsPage from "./components/user/UserSecuritySettingsPage"
import UserAppearanceSettingsPage from "./components/user/UserAppearanceSettingsPage"
import UserMarkingTablesSettingsPage from "./components/user/UserMarkingTablesSettingsPage"
import GroupSettingsPage from "./components/group/GroupSettingsPage"
import GroupUserProfilePage from "./components/group/GroupUserProfilePage"
import HardCodedExamplePage from "./components/solution/HardCodedExamplePage";
import {ApplicationStateInterface} from "./types/ApplicationStateInterface";
import {ActionType} from "./types/ActionType";

function App() {
  useEffect(() => {
    if (localStorage.getItem("token") && localStorage.getItem("id")) {
      localStorage.removeItem("token")
      localStorage.removeItem("id")
    }
  }, [])
  const initialState: ApplicationStateInterface = {
    loggedIn: Boolean(localStorage.getItem("token")),
    homePageIn: true,
    userState: {
      token: localStorage.getItem("token")!,
      userId: localStorage.getItem("id")!,
    },
  }
  const [role, setRole] = useState<string>("")
  function ourReducer(draft: ApplicationStateInterface, action: ActionType) {
    switch (action.type) {
      case "login":
        draft.loggedIn = true
        draft.userState = action.data
        break
      case "logout":
        draft.loggedIn = false
        break
      case "homePageIn":
        draft.homePageIn = true
        break
      case "homePageOut":
        draft.homePageIn = false
        break
    }
  }
  const [state, dispatch] = useImmerReducer<ApplicationStateInterface, ActionType>(
    ourReducer,
    initialState
  )

  useEffect(() => {
    if (state.loggedIn) {
      localStorage.setItem("token", state.userState.token)
      localStorage.setItem("id", state.userState.userId)
    } else {
      localStorage.removeItem("token")
      localStorage.removeItem("id")
    }
  }, [state])

  return (
    <UserContext.Provider value={state}>
      <DispatchContext.Provider value={dispatch}>
        <GroupRoleContext.Provider value={{ role }}>
          <GroupSetRoleContext.Provider value={{ setRole }}>
            <BrowserRouter>
              <Header />
              <Routes>
                <Route
                  path="/"
                  element={state.loggedIn ? <HomePage /> : <HomeGuestPage />}
                />
                <Route
                  path="/:id/assignments"
                  element={<AssignmentsStudentDisplayedPage />}
                />
                <Route path="/settings" element={<UserSettingsPage />} />
                <Route
                  path="/settings/general"
                  element={<UserGeneralSettingsPage />}
                />
                <Route
                  path="/settings/security"
                  element={<UserSecuritySettingsPage />}
                />
                <Route
                  path="/settings/appearance"
                  element={<UserAppearanceSettingsPage />}
                />
                <Route
                  path="/settings/markingTables"
                  element={<UserMarkingTablesSettingsPage />}
                />
                <Route path="/login" element={<LoginPage />} />
                <Route path="/register" element={<RegisterPage />} />
                <Route path="/create/group" element={<GroupCreatePage />} />
                <Route path="/group/:id" element={<GroupPage />}>
                  <Route path="settings" element={<GroupSettingsPage />} />
                  <Route path="users" element={<GroupUsersPage />} />
                  <Route
                    path="userProfileInGroup/:userProfileId"
                    element={<GroupUserProfilePage />}
                  />
                  <Route
                    path="assignments"
                    element={<AssignmentsGroupDisplayedPage />}
                  />
                  <Route
                    path="assignments/done"
                    element={<AssignmentsTypesPage type={"done"} />}
                  />
                  <Route
                    path="assignments/expired"
                    element={<AssignmentsTypesPage type={"expired"} />}
                  />
                  <Route
                    path="assignments/todo"
                    element={<AssignmentsTypesPage type={"todo"} />}
                  />
                  <Route path="assignments/add" element={<AddAssigmentPage />} />
                  <Route
                    path="assigment/:idAssigment"
                    element={<AssigmentSpecPage />}
                  />
                  <Route
                    path="solution/:idUser/:idAssigment"
                    element={<SolutionPage />}/>
                  <Route
                    path="solutions/uncheck"
                    element={<SolutionsTypesPage type={"uncheck"} />}
                  />
                  <Route
                    path="solutions/late"
                    element={<SolutionsTypesPage type={"late"} />}
                  />
                  <Route
                    path="solutions/check"
                    element={<SolutionsTypesPage type={"check"} />}
                  />


                  <Route path="*" element={<NotFoundPage />} />
                </Route>
                <Route path="/group/:id/solution/:idUser/:idAssigment/example" element={<HardCodedExamplePage/>}/>
                <Route path="*" element={<NotFoundPage />} />
              </Routes>
            </BrowserRouter>
          </GroupSetRoleContext.Provider>
        </GroupRoleContext.Provider>
      </DispatchContext.Provider>
    </UserContext.Provider>
  )
}

export default App
