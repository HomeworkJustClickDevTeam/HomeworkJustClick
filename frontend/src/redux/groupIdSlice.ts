import {createSlice, PayloadAction} from "@reduxjs/toolkit";
import {RootState} from "./store";

type GroupIdSliceType = number | null
const initialState = null as GroupIdSliceType

export const groupIdSlice = createSlice({
    name: 'groupId',
    initialState,
    reducers: {
        setGroupId: (state, action: PayloadAction<GroupIdSliceType>) => action.payload
    }
})

export const {setGroupId} = groupIdSlice.actions
export const selectGroupId = (state: RootState) => state.groupId
export default groupIdSlice.reducer