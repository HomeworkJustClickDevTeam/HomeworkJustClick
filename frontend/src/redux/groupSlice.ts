import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import { GroupInterface } from "../types/GroupInterface"
import { RootState } from "./store"

type GroupSliceType = GroupInterface | null
const initialState = null as GroupSliceType
export const groupSlice = createSlice({
  name:'group',
  initialState,
  reducers:{
    setGroup:(state, action:PayloadAction<GroupSliceType>) => action.payload
  }
})

export const {setGroup} = groupSlice.actions
export const selectGroup = (state:RootState) => state.group
export default groupSlice.reducer