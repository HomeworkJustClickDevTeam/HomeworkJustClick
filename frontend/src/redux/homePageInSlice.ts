import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import { RootState } from "./store"

const initialState: boolean = true
export const homePageInSlice = createSlice({
  name: 'homePageIn',
  initialState,
  reducers:{
    setHomePageIn:(state, action:PayloadAction<boolean>) => action.payload
  }
})

export const {setHomePageIn} = homePageInSlice.actions
export const selectHomePageIn = (state:RootState) => state.homePageIn
export default homePageInSlice.reducer