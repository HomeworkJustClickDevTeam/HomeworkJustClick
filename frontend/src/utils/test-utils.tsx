import {configureStore, PreloadedState} from "@reduxjs/toolkit";
import userStateReducer from "../redux/userStateSlice";
import roleReducer from "../redux/roleSlice";
import groupReducer from "../redux/groupSlice";
import isLoadingReducer from "../redux/isLoadingSlice";
import homePageInReducer from "../redux/homePageInSlice";
import {Provider} from "react-redux";
import {render, RenderOptions} from "@testing-library/react";
import {RootState, setupStore} from "../redux/store";
import {ToolkitStore} from "@reduxjs/toolkit/dist/configureStore";
import {PropsWithChildren} from "react";

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: ToolkitStore
}
export const renderWithProviders = (
    ui: React.ReactElement,
    {
      preloadedState = {},
      store = setupStore(preloadedState),
      ...renderOptions
    }: ExtendedRenderOptions={})=> {

  const Wrapper = ({children}:PropsWithChildren<{}>):JSX.Element => {
    return <Provider store={store}>{children}</Provider>
  }

  return {store, ...render(ui, {wrapper: Wrapper, ...renderOptions})}
}