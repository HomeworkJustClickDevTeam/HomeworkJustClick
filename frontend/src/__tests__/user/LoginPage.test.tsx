import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { setupStore } from "../../redux/store";
import LoginPage from "../../components/user/LoginPage";
import { loginPageHandler } from "../../mocks/handlers";
import { renderWithProviders } from "../../utils/test-utils";
import { fireEvent, screen } from "@testing-library/react";
import { BrowserRouter, MemoryRouter, Route, Routes } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { setupServer } from "msw/node";
import '@testing-library/jest-dom';
import React from "react";
import App from "../../App";
const server = setupServer(...loginPageHandler)
beforeAll(() => server.listen())
beforeEach(()=> {
  jest.clearAllMocks()
})
afterEach(()=> server.resetHandlers())
afterAll(() => server.close())


describe("check if user can login with page navigation", ()=> {
  it("checks if login page renders correctly", () => {
    const renderedHomeGuestPage = renderer
      .create(
        <Provider store={setupStore()}>
          <BrowserRouter>
            <LoginPage></LoginPage>
          </BrowserRouter>
        </Provider>
      )
    expect(renderedHomeGuestPage).toMatchSnapshot()
  });

  it("checks if user can login", async ()=>{
    const user = userEvent.setup()
    renderWithProviders(
      <MemoryRouter initialEntries ={['/login']}>
        <App/>
      </MemoryRouter>
    )
    const registerButton = screen.getByRole('button', {name: /zarejestruj się/i})
    const loginButton = screen.getByRole('button', {name: /zaloguj się/i})
    const emailTextBox = screen.getByPlaceholderText(/Email/i)
    const passwordTextBox = screen.getByPlaceholderText(/hasło/i)
    expect(registerButton).toBeEnabled()
    expect(loginButton).toBeEnabled()
    expect(screen.getByRole('heading',{name: /zaloguj się/i})).toBeInTheDocument()
    expect(emailTextBox).toBeInTheDocument()
    expect(passwordTextBox).toBeInTheDocument()
    expect(passwordTextBox).toBeEmptyDOMElement()
    expect(emailTextBox).toBeEmptyDOMElement()
    await user.type(emailTextBox, "adam_nowak@gmail.com")
    expect(emailTextBox).toHaveValue("adam_nowak@gmail.com")
    await user.type(passwordTextBox, "123")
    expect(passwordTextBox).not.toHaveValue("")
    await user.click(loginButton)
    const pageNavigation = await screen.findAllByText(/moje grupy/i)
    expect(pageNavigation).toBeInstanceOf(Array)
    expect(pageNavigation).toHaveLength(2)
  })

})

