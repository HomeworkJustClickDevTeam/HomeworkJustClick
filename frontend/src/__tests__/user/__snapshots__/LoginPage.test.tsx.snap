// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check if user can login with page navigation checks if login page renders correctly 1`] = `
<div
  className="flex w-screen flex-col items-center text-center font-lato text-sm select-none"
  data-testid="login-page"
>
  <img
    alt="Kółko po lewej stronie"
    className="fixed left-[7%]  bottom-[3%] scale-50 xl:scale-100 -z-50"
    src={
      {
        "process": [Function],
      }
    }
  />
  <img
    alt="Kółko po prawej stronie"
    className="fixed right-[9%] top-[18%] scale-50 translate-x-[25%] xl:transform-none -z-50"
    src={
      {
        "process": [Function],
      }
    }
  />
  <h1
    className="mt-32 mb-16 text-6xl"
  >
    Zaloguj się
  </h1>
  <form
    className="flex flex-col items-center"
    onSubmit={[Function]}
  >
    <input
      className="border-b-2 border-b-light_gray mb-8 text-center placeholder:text-light_gray placeholder:text-[12px] w-60 xl:text-[20px] xl:placeholder:text-[20px] xl:w-[320px]"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
    />
    <input
      className="border-b-2 border-b-light_gray mb-12 text-center placeholder:text-light_gray placeholder:text-[12px] w-60 xl:text-[20px] xl:placeholder:text-[20px] xl:w-[320px]"
      name="password"
      onChange={[Function]}
      placeholder="Hasło"
      type="password"
    />
    <button
      className=" border-main_blue w-40 h-8 rounded-[5px] border-2 xl:w-[288px] xl:h-[56px] xl:border-[3px] xl:rounded-[10px] text-main_blue xl:text-[28px] hover:bg-hover_gray hover:shadow-md active:bg-opacity-60"
      type="submit"
    >
      Zaloguj się
    </button>
    <p
      className="text-[8px] xl:text-base text-main_blue"
    >
      Nie pamiętasz hasła?
    </p>
    <p
      className="text-[8px] xl:text-base mt-4"
    >
      Nie masz jeszcze konta?
    </p>
    <a
      href="/register"
      onClick={[Function]}
    >
      <button
        className=" bg-main_blue w-40 h-8 rounded-[5px] xl:rounded-[10px] xl:w-[288px] xl:h-[56px] text-white xl:text-[28px] hover:bg-hover_blue hover:shadow-md active:shadow-none"
        type="button"
      >
        Zarejestruj się
      </button>
    </a>
  </form>
</div>
`;
